File: apps/android/app/src/main/java/com/cursumi/app/core/model/SocialModels.kt
package com.cursumi.app.core.model

import kotlinx.serialization.Serializable

// Chat, notas, reflexiones, reseñas, referidos, blog, materiales y juegos en vivo.

@Serializable data class ChatSender(val name: String? = null)
@Serializable data class ChatMessage(val id: String, val body: String, val senderId: String = "", val createdAt: String = "", val sender: ChatSender? = null)
@Serializable data class Conversation(val id: String, val messages: List<ChatMessage>? = null)

@Serializable data class NoteRef(val id: String, val title: String)
@Serializable data class Note(val id: String, val content: String, val createdAt: String = "", val course: NoteRef? = null, val lesson: NoteRef? = null)

@Serializable data class NamedUser(val name: String? = null)
@Serializable data class Reflection(val id: String, val content: String, val user: NamedUser? = null)
@Serializable data class Review(val id: String, val rating: Int = 0, val comment: String? = null, val user: NamedUser? = null)
@Serializable data class ReviewsResponse(val reviews: List<Review> = emptyList(), val average: Double? = null, val total: Int? = null)

@Serializable
data class Referral(
    val referralCode: String? = null,
    val referralLink: String = "",
    val totalReferrals: Int = 0,
    val pendingReferrals: Int = 0,
    val earnedReferrals: Int = 0,
    val totalEarnedCents: Int = 0,
    val totalPaidCents: Int = 0,
)

@Serializable data class BlogPostSummary(val title: String, val slug: String, val excerpt: String? = null, val coverImageUrl: String? = null)
@Serializable data class BlogAuthor(val name: String? = null)
@Serializable data class BlogPost(val title: String, val slug: String = "", val content: String = "", val coverImageUrl: String? = null, val author: BlogAuthor? = null)

@Serializable data class OrgMaterial(val id: String, val name: String, val description: String? = null, val fileUrl: String, val fileType: String = "")
@Serializable data class OrgMaterialsResponse(val orgName: String? = null, val materials: List<OrgMaterial> = emptyList())

// ── Juegos en vivo ──────────────────────────────────────────────────────────

@Serializable data class GameParticipant(val id: String, val nickname: String, val score: Int = 0)
@Serializable data class GameQuestion(val id: String, val question: String, val options: List<String> = emptyList())
@Serializable data class GameInfo(val id: String, val status: String, val currentQuestion: Int? = null, val code: String? = null, val questions: List<GameQuestion>? = null)
@Serializable data class MyAnswer(val selectedOption: Int? = null)

@Serializable
data class GameState(
    val game: GameInfo,
    val currentQ: GameQuestion? = null,
    val participants: List<GameParticipant> = emptyList(),
    val myAnswer: MyAnswer? = null,
    val myParticipantId: String? = null,
    val myNickname: String? = null,
) {
    val ranked get() = participants.sortedByDescending { it.score }
}

File: apps/ios/Cursumi/Core/Models/SocialModels.swift
import Foundation

// Chat, notas, reflexiones, reseñas, referidos, blog, materiales y juegos en vivo.

struct ChatMessage: Decodable, Identifiable, Equatable {
    struct Sender: Decodable, Equatable { let name: String? }
    let id: String
    let body: String
    let senderId: String
    let createdAt: Date
    let sender: Sender?
}

struct Conversation: Decodable {
    let id: String
    let messages: [ChatMessage]?
}

struct Note: Decodable, Identifiable, Equatable {
    struct Ref: Decodable, Equatable { let id: String; let title: String }
    let id: String
    let content: String
    let createdAt: Date
    let course: Ref?
    let lesson: Ref?
}

struct Reflection: Decodable, Identifiable, Equatable {
    struct User: Decodable, Equatable { let name: String? }
    let id: String
    let content: String
    let user: User?
}

struct Review: Decodable, Identifiable, Equatable {
    struct User: Decodable, Equatable { let name: String? }
    let id: String
    let rating: Int
    let comment: String?
    let user: User?
}

struct ReviewsResponse: Decodable {
    let reviews: [Review]
    let average: Double?
    let total: Int?
}

struct Referral: Decodable, Equatable {
    let referralCode: String?
    let referralLink: String
    let totalReferrals: Int
    let pendingReferrals: Int
    let earnedReferrals: Int
    let totalEarnedCents: Int
    let totalPaidCents: Int
}

struct BlogPostSummary: Decodable, Identifiable, Equatable {
    var id: String { slug }
    let title: String
    let slug: String
    let excerpt: String?
    let coverImageUrl: String?
}

struct BlogPost: Decodable {
    struct Author: Decodable { let name: String? }
    let title: String
    let slug: String
    let content: String
    let coverImageUrl: String?
    let author: Author?
}

/// `/api/blog` devuelve array o `{ posts }`.
struct BlogListResponse: Decodable {
    let posts: [BlogPostSummary]
    init(from decoder: Decoder) throws {
        if let list = try? decoder.singleValueContainer().decode([BlogPostSummary].self) {
            posts = list
        } else {
            let c = try decoder.container(keyedBy: Keys.self)
            posts = (try? c.decode([BlogPostSummary].self, forKey: .posts)) ?? []
        }
    }
    private enum Keys: String, CodingKey { case posts }
}

struct OrgMaterial: Decodable, Identifiable, Equatable {
    let id: String
    let name: String
    let description: String?
    let fileUrl: String
    let fileType: String
}

struct OrgMaterialsResponse: Decodable {
    let orgName: String?
    let materials: [OrgMaterial]
}

// MARK: - Juegos en vivo (lado jugador)

struct GameParticipant: Decodable, Identifiable, Equatable {
    let id: String
    let nickname: String
    let score: Int
}

struct GameQuestion: Decodable, Identifiable, Equatable {
    let id: String
    let question: String
    let options: [String]
}

struct GameState: Decodable {
    struct Game: Decodable {
        let id: String
        let status: String // waiting | active | finished
        let currentQuestion: Int?
    }
    struct MyAnswer: Decodable { let selectedOption: Int? }
    let game: Game
    let currentQ: GameQuestion?
    let participants: [GameParticipant]
    let myAnswer: MyAnswer?
    let myParticipantId: String?
    let myNickname: String?

    var ranked: [GameParticipant] { participants.sorted { $0.score > $1.score } }
}
